### PR TITLE
Add resolving enviroment variables in Pipfile URL.

### PR DESCRIPTION
**Pull request by Misoslav**
## Related Issues and Dependencies

Fixes: #158 

## This introduces a breaking change

- [ ] Yes
- [x] No

## This Pull Request implements

injecting creditials into pipfiles via enviromental variables

### CODE REVIEW

File: micropipenv.py
Comment (fridex) [inline micropipenv.py:1004]: Do we want to have an option to control injecting environment variables using a micropipenv specific environment variable? I could imagine a use-case when micropipenv would be used for installing dependencies in a CI system and an attacker could adjust files to reveal sensitive environment variables to CI logs.

File: micropipenv.py
Comment (fridex) [inline micropipenv.py:978]: The function can be private:

```suggestion
def _resolve_nested_variables(url):
```

File: micropipenv.py
Comment (frenzymadness) [reply micropipenv.py:1004]: Interesting point of view. Do you mean something like `MICROPIPENV_ENABLE_ENV_VARS`? And only if the mentioned variable is defined this function would do something with the variables in the files?

I must admit that it sounds reasonable but also kinda complicated for a user. Do you have an example of a tool doing something similar?

Comment (fridex) [reply micropipenv.py:1004]: > Interesting point of view. Do you mean something like `MICROPIPENV_ENABLE_ENV_VARS`? And only if the mentioned variable is defined this function would do something with the variables in the files?

Yes, in that case, the software could be more secure (inverted logic could also work fine but would require a proper setup by the CI system). 

> I must admit that it sounds reasonable but also kinda complicated for a user. Do you have an example of a tool doing something similar?

None I could think of from the top of my head. Just an idea during the review.

Comment (frenzymadness) [reply micropipenv.py]: I'm still thinking about how the additional option to enable/disable the feature can make it more secure. Let's say I'm an attacker and my plan is to modify Pipfile or Pipfile.lock to show me some secrets in CI logs. If I'm able to do that, I'm also able to modify the rest of the application files (because they are usually sent from the same place together) and use another way to uncover the secrets or adjust settings to enable the feature we are talking about.

I'll try to discuss it with my team today. They might have some experience with that.

Comment (frenzymadness) [reply micropipenv.py]: One more note from our morning brainstorming: If a user uses a CI/CD system for an open-source project, they should make sure that secrets in these services are available only for a run from protected branches because everybody can open a PR for your project and a random PR should not have access to any secrets because only some deploy phase needs them.

If this is true and the CI system is properly configured it would need an attacker's change to go through review to get it to a protected branch to have any impact. And in a case like that, an attacker can also modify other files, enable other insecure features and do a lot of stuff we cannot handle and it's the responsibility of a maintainer to take care of that.

**Comment by goern** [comment]
@fridex @frenzymadness shall I add Misoslav as a collaborator?

**Comment by frenzymadness** [comment]
> @fridex @frenzymadness shall I add Misoslav as a collaborator?

No need to do it now, we'll see whether he will want to continue contributing or not.

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:213]: There is no handling of the exception this call should raise. If this test passes, there is something wrong.

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:209]:
```suggestion
def test_install_pipenv_env_vars_undefined(venv):
```

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:200]: Could you please adjust the doc strings to represent the specific test?

**Review by frenzymadness — CHANGES_REQUESTED** [review]
A few comments from me. Will see what the tests will produce.

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:300]: Unrelated change:
```suggestion
```

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:203]: `+` is not supported for dicts, take a look at the `.update` method.

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:220]:
```suggestion
    """Test installation using default values of environment variables."""
```

File: tests/test_micropipenv.py
Comment (frenzymadness) [inline tests/test_micropipenv.py:223]: Either you use `subprocess.run` with `check=True` or `subprocess.check_call`. Here you combined both approaches. And because we don't expect the command to fail here, you can also just use `subprocess.run` without `check` because no matter what exit code the command has, we check the installed packages.

File: micropipenv.py
Comment (fridex) [reply micropipenv.py]: Yes, you are right. Let's keep this configuration on CI side (similarly as `/ok-to-test`). Sorry for bringing confusion. 👍🏻

**Comment by frenzymadness** [comment]
I'm gonna take over this effort before I lost the context of it. I believe I know what has to be done to finish it. There are a few new commits from me that should fix the tests and make this PR ready for review. We should squash all of them to one before or during the merge.

@Misoslav thanks for the contribution. If you are interested in what I had to do to finish the PR take a look at the last commits and do not hesitate to ask.

**Review by fridex — COMMENTED** [review]
Very nice, changes look good to me. Thanks! 💯 

Would it be possible to squash commits? 11 commits sound like a lot for this feature.

**Comment by frenzymadness** [comment]
You can use "Squash and merge" on Github to squash all of the commits to one.

**Comment by fridex** [comment]
> You can use "Squash and merge" on Github to squash all of the commits to one.

OK, can do it. Wanted to use our bots for that.

**Pull request merged**
